### PR TITLE
Code Insights: Fix funky colors of insight data-series

### PR DESCRIPTION
**Pull request by vovakulikov**
Closes https://github.com/sourcegraph/sourcegraph/issues/24683

## Context

On main at the moment we have a problem with chart when two or more points don't have values for some time-x value when other point/points have values at this x point

```
[
  { x: 1588965700286 - 4 * 24 * 60 * 60 * 1000, a: null, b: null },
  { x: 1588965700286 - 2 * 24 * 60 * 60 * 1000, a: 94, b: 200 },
  // Datum below doesn't have b value but have a value  
  { x: 1588965700286 - 1.5 * 24 * 60 * 60 * 1000, a: 134, b: null },
    // Datum below doesn't have a value but have b value 
  { x: 1588965700286 - 1.3 * 24 * 60 * 60 * 1000, a: null, b: 150 },
  { x: 1588965700286 - 1 * 24 * 60 * 60 * 1000, a: 134, b: 190 },
]
```

So because our assumption about this case was wrong in the line chart code base (initially we thought that all points will have some values at some point at the time X-axis or neither of points has that value and therefore we don't have this point on the x-axis in the datum array at all).

But it turned out that this is the case cause for BE insights it's possible to have some values for one series and doesn't have value for other series at one particular point on the x-axis. 

As a result, we have this visual behavior 

image

## Solution

This Pr adds the preparation step for the line chart data (datum list) and series configs. 
We convert this data
```
[
  { x: 1588965700286 - 4 * 24 * 60 * 60 * 1000, a: null, b: null },
  { x: 1588965700286 - 2 * 24 * 60 * 60 * 1000, a: 94, b: 200 },
  // Datum below doesn't have b value but have a value  
  { x: 1588965700286 - 1.5 * 24 * 60 * 60 * 1000, a: 134, b: null },
    // Datum below doesn't have a value but have b value 
  { x: 1588965700286 - 1.3 * 24 * 60 * 60 * 1000, a: null, b: 150 },
  { x: 1588965700286 - 1 * 24 * 60 * 60 * 1000, a: 134, b: 190 },
],
series: [
 { dataKey: 'a', name: 'A metric' },
 { dataKey: 'b', name: 'B metric' },
],
```

into this
```
series: [
  { dataKey: 'a', name: 'A metric', points: [ {x, y}, {x,y} ... ] },
  { dataKey: 'b', name: 'B metric', points: [ {x, y}, {x,y} ... ] },
]
```

Each series now has its own datum list (points array) we process this array and remove null/undefined gaps between points but preserve nullish data at the beginning for the non-existent points patter (line background)
Also because in this case, not all series may have all points at some x-axis value it's possible to not have some chart value in tooltip info. See screenshots below

<table>
 <tr><th>With a and b series values</th><th>With one series missing value</th></tr>
 <tr><td>
 
Screenshot 2021-09-15 at 19 22 17

 </td>
 <td>
 
image

 </td>
 </tr>
</table>

## Work in progress
- Add unit tests to the `getProcessedChartData` function
- Probably change other logic for data preparation and move away from datum reach list to the series datum instead.
- Create a separate story for this case

### CODE REVIEW

**Comment by vovakulikov** [comment]
@coury-clark @CristinaBirkel Also curious what do you think about our data structures for the charts. I see points in having reach datum array (with all y points in one map/object) but like this PR shows this structure has its disadvantages. Do you think we could change something in our chart data shape on the BE to avoid this FE processing.

**Comment by coury-clark** [comment]
> @coury-clark @CristinaBirkel Also curious what do you think about our data structures for the charts. I see points in having reach datum array (with all y points in one map/object) but like this PR shows this structure has its disadvantages. Do you think we could change something in our chart data shape on the BE to avoid this FE processing.

I'm not fully sure I understand everything in this PR but let me see if I can separate a few things. Correct me if I'm wrong on any of these.
1. The way frontend insights generate data is such that all of the underlying series will have the same set of distinct time points.
2. The backend generates insight series separately and will try to reuse previously generated series if possible. This results in potentially disjoint sets of time points.
3. Something about the way our chart library renders these series causes a problem if each series has different x-values (is this correct, or is it something on the frontend in our code that is causing the problem?)

In general, I'd say I don't think the behavior of returning the underlying series as they exist is a problem. I am not favorable to the idea that we would manipulate the series on a per-chart basis in the API - I think that it's up to the consumers of this data to decide how it needs to be rendered, otherwise we would end up having to build bespoke APIs for each need.

To be clear, the API as-is returns an array of (time, value) tuples per **series**. I'm not entirely sure what you mean by a data structure having all y-values in a single object otherwise.

It can help to think of the underlying as "points in time for which a value was recorded", and with that in mind, it becomes more intuitive to render the series without manipulation.

File: client/web/src/views/components/content/view-content/chart-view-content/charts/line/helpers/get-processed-chart-data.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    // Preserve null values at the beginning of the series data list
+    // but remove null holes between the points further.
+    const nullBeginningValues = data.slice(0, firstNonNullablePointIndex)
+    const pointsWithoutHoles = data
+        // Get values after null area
+        .slice(firstNonNullablePointIndex)
+        .filter(point => point.y !== null)
+
+    return [...nullBeginningValues, ...pointsWithoutHoles]
```
Comment (valerybugakov) [inline client/web/src/views/components/content/view-content/chart-view-content/charts/line/helpers/get-processed-chart-data.ts:102]: Would adding the index condition to filter work here?

```suggestion
    // Preserve null values at the beginning of the series data list
    // but remove null holes between the points further.
    return data.filter((point, index) => point.y !== null || index <= firstNonNullablePointIndex)
```

File: client/web/src/views/components/content/view-content/chart-view-content/charts/line/helpers/get-processed-chart-data.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ * ```
+ * data: [
+ *   { a: 2, b: 3, x: 2},
+ *   { a: 4, b: 5, x: 3},
+ *   { a: null, b: 6, x: 4}
+ * ] → series: [
+ *     { name: a, data: [{ y: 2, x: 2 }, { y: 4, x: 3 }],
+ *     { name: b, data: [{ y: 3, x: 2 }, { y: 5, x: 3 }, { y: 6, x: 4 }]
+ *   ]
+ * ```
```
Comment (valerybugakov) [inline client/web/src/views/components/content/view-content/chart-view-content/charts/line/helpers/get-processed-chart-data.ts:39]: This is super helpful

**Review by valerybugakov — APPROVED** [review]
✅

File: client/web/src/views/components/content/view-content/chart-view-content/charts/line/helpers/get-processed-chart-data.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ * │░░░░░░░░░░░░░░░                     │       │░░░░░░░░░░░░░░░                     │
+ * │░░░░░░░░░░░░░░░                     │       │░░░░░░░░░░░░░░░                     │
+ * │░░░░░░░░░░░░░░░                ■    │       │░░░░░░░░░░░░░░░                ■    │
+ * │░░░░░░░░░░░░▽░░    ■                │       │░░░░░░░░░░░░▽░░    ■                │
+ * │░░░░░░░░░░░░░░░          ▽          │──────▶│░░░░░░░░░░░░░░░                     │
+ * │░░░░░░▽░░░░░░░░ ■                   │       │░░░░░░▽░░░░░░░░ ■                   │
+ * │░░░░░░░░░░░░░░░       ■             │       │░░░░░░░░░░░░░░░       ■             │
+ * │░░░▽░░░░░░░░░░░                     │       │░░░▽░░░░░░░░░░░                     │
+ * │░░░░░░░░░░░░░░░             ▽       │       │░░░░░░░░░░░░░░░                     │
+ * └────────────────────────────────────┘       └────────────────────────────────────┘
```
[the remark this answers is not in the document]
Comment (tjkandala) [reply client/web/src/views/components/content/view-content/chart-view-content/charts/line/helpers/get-processed-chart-data.ts:84]: @vovakulikov do you do this by hand? so pretty

[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/web/src/views/components/content/view-content/chart-view-content/charts/line/helpers/get-processed-chart-data.ts:85]: @tjkandala I used to do this by hand but recently I've found https://monodraw.helftone.com This chart is built by this thing

**Comment by vovakulikov** [comment]
@coury-clark You're right. I was wrong about our API. It's actually our FE does these merging from multiple series one big object with all y-values from all data series. The bad part that's with this PR we do the right opposite thing by on low-level components. But unfortunately, we can't go away from it cause we still need to support old extensions like API cause some external code insights like extensions like codecov or snyk are using this card as well, and migrate data shape for our insights would mean that we have to change data API in extension as well. 

Sorry to bother you.

**Pull request merged**